### DISCUSSION TITLE
setProgressHandler always returns the entire phrase

### DISCUSSION THREAD

**Discussion by adamcastle** [dlutton/flutter_tts#193] [Q&A] [answered]
Hi,
Awesome plugin but  I'm having trouble trying to get the setProgressHandler to return a single/current word.
Currently the variable "word" and "text" always return the entire phrase.  I have also added the handler to your example without any luck. I have also noticed the event is only ever executed once.  

`flutterTts.speak("this is a test");`
`flutterTts.setProgressHandler((String text, int startOffset, int endOffset, String word) {`
`//word = "this is a test";`
`});`

Any ideas ?

Thanks,
Adam.

**Comment by dlutton** [comment]
What are you trying to accomplish exactly with `setProgressHandler`?

**Reply by adamcastle** [reply]
Thanks for the reply,

I was hoping to get the current word so I can highlight it within the phrase on the screen. So you can following the talking on the screen.

So I would end up with something like the following.

"**This** is a test"
"This **is** a test"
"This is **a** test"
"This is a **test**"

Hope that makes sense.

Cheers,
Adam.

**Comment by adamcastle** [comment]
Hi,
I have created an example of what I am trying to achieve below.
https://k2nzfiles.blob.core.windows.net/temp/Speech.html

**Reply by dlutton** [reply]
Sure that make sense, sorry I've been extremely busy, but I'll try to get a working example for you soon.

**Comment by adamcastle** [comment]
That would be fantastic and it’s all good, I’m also flat out. Thanks for all your hard work!

**Comment by dlutton** [comment, marked as answer]
Okay I apologize it took so long, I quickly threw something together that matches the example in your link based on my example in this repo.
https://gist.github.com/dlutton/30b065449cb2eb941dfd6bef86aeefaf

Let me know if you have any questions.  The widget that's doing most of the work is `_textFromInput`.  Hopefully this helps.

**Reply by adamcastle** [reply]
That's awesome! Thanks heaps.

**Reply by dlutton** [reply]
Should I close this discussion, or do you need help with anything else related to this?

**Reply by adamcastle** [reply]
Hi, 

Thanks for all your help on this. Unfortunately this no longer works. I think maybe because I upgraded flutter but are not sure. I have created a few new projects from scratch and still have the same issue. I have tried flutter_tts version 2.1.2 and 3. 

Are you able to add this code into your main example. 

Do you want me to raise a bug on this ?

Cheers,
Adam

**Reply by dlutton** [reply]
You just need to add in the null safety options.  I have updated the gist so that it will work with the latest flutter and flutter_tts version.

**Reply by adamcastle** [reply]
Hey, 

Thanks for the reply. 

Below is the Flutter and Dart version that I am using.

Flutter 2.0.4 • channel stable • https://github.com/flutter/flutter.git
Framework • revision b1395592de (6 days ago) • 2021-04-01 14:25:01 -0700
Engine • revision 2dce47073a
Tools • Dart 2.12.2

I can build and run the example on Chrome and Android but i'm getting the same issue.

I presume I have enabled null safety options because it builds and runs.

Any ideas?  Sorry to be a pain.

Cheers,
Adam.

**Reply by dlutton** [reply]
no worries, you're not being a pain at all.  what's the issue you're encountering?

**Reply by adamcastle** [reply]
Hey,

Its not highlighting the individual words, only the entire sentence. You can see the example below .

![image](https://user-images.githubusercontent.com/3186433/113945521-92110680-985a-11eb-971d-15719f8d633f.png)

In fact if you look at the code below the event isn't being fired in this case. Debugging on Chrome

![image](https://user-images.githubusercontent.com/3186433/113945920-722e1280-985b-11eb-8b78-db0bf013206d.png)

When debugging on my Android tablet the event fires once.
![image](https://user-images.githubusercontent.com/3186433/113946442-94746000-985c-11eb-94d5-6ab76cbf0e20.png)

What do you think?

Thanks again!

**Reply by dlutton** [reply]
this isn't supported in web. I haven't tested this with an Android tablet.  I'll have to see if I can replicate it with an emulator.  Have you tested on an android phone and/or iphone?

**Reply by duy1998** [reply]
It is wrong in Android real devices with long text. It seems it does not color sequentially the words being read.
Anyone see it ?
